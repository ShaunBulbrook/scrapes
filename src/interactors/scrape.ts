const Cheerio = require("cheerio");

/**
 * Takes markup and a query selector. Returns the text under elements that match the selector
 * @param markup string of html
 * @param selector query selector to rip text from
 */
module.exports.scrapeSelection = (markup: string, selector: string): string => {
	const $: any = Cheerio.load(markup);
	return $(selector).text() || "No results found.";
};

/**
 * Takes markup and returns an array of all links of the page.
 * @param markup string of html
 */
module.exports.scrapeLinks = (markup: string): string[] => {
	const $: any = Cheerio.load(markup);
	const links: string[] = [];

	$("a[href]").each((index: number, element: any) =>
		links.push($(element).attr("href"))
	);

	if (links.length > 0) {
		return links;
	} else {
		throw "No links could be found.";
	}
};
