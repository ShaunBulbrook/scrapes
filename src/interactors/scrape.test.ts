import { scrapeSelection, scrapeLinks } from "./scrape";
import { markup, hackerNewsMarkup } from "../stubs/markup";

const fetch = jest.fn();
fetch.mockReturnValue(markup);

describe("scrapeSelection", () => {
	it("should inform the user that if it failed to find results", () => {
		expect(scrapeSelection("", "")).toBe("No results found.");
	});
	it("should return the content of a targeted element", () => {
		expect(scrapeSelection(markup, "h1")).toBe("Example Domain");
	});
	it("should return the content of multiple targeted elements", () => {
		expect(scrapeSelection(markup, "p")).toMatchSnapshot("paragraphs");
	});
	it("should be able to use attribute selectors", () => {
		expect(
			scrapeSelection(markup, `a[href="http://www.iana.org/domains/example"]`)
		).toBe("More information...");
	});
});

describe("scrapeLinks", () => {
	it("should throw an error if no links are found", () => {
		function scrapeLinksFromNothing() {
			scrapeLinks("");
		}
		expect(scrapeLinksFromNothing).toThrowError("No links could be found.");
	});
	it("should return a link on a page", () => {
		expect(scrapeLinks(markup)[0]).toBe("http://www.iana.org/domains/example");
	});
	it("should return different links", () => {
		expect(scrapeLinks(hackerNewsMarkup)).toContain(
			"http://www.antenna-theory.com/"
		);
	});
});
