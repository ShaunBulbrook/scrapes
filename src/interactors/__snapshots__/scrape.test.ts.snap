// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrapeSelection should return the content of multiple targeted elements: paragraphs 1`] = `
Array [
  "
				This domain is established to be used for illustrative examples in
				documents. You may use this domain in examples without prior
				coordination or asking for permission.
			",
  "
				More information...
			",
]
`;
